mpcs/src/basefold/commit_phase.rs

```diff
@@ -201,54 +326,139 @@ where
-                lo + (hi - lo) * challenge
+    debug_assert_eq!(folding_coeffs.len(), 1 << level);
+
+    // take codewords match with target length then fold
+    let codewords_matched =
+        pop_front_while(codewords, |codeword| codeword.values.len() == target_len);
+    // take codewords match next target length in preparation of being committed together
+    let codewords_next_level_matched = pop_front_while(codewords, |codeword| {
+        codeword.values.len() == next_level_target_len
+    });
+
+    // optimize for single codeword match
+    let folded_codeword = if (usize::from(running_codeword_opt.is_some()) + codewords_matched.len())
+        == 1
+        && codewords_next_level_matched.is_empty()
+    {
+        RowMajorMatrix::new(
+            running_codeword_opt
+                .or_else(|| codewords_matched.first().map(|m| m.as_view()))
+                .unwrap()
+                .values
+                .par_chunks_exact(2)
+                .zip(folding_coeffs)
+                .map(|(ys, coeff)| codeword_fold_with_challenge(ys, challenge, *coeff, inv_2))
+                .collect::<Vec<_>>(),
+            2,
+        )
+    } else {
+        // aggregate codeword with same length
+        let codeword_to_fold = (0..target_len)
+            .into_par_iter()
+            .map(|index| {
+                running_codeword_opt
+                    .into_iter()
+                    .chain(codewords_matched.iter().map(|m| m.as_view()))
+                    .map(|codeword| codeword.values[index])
+                    .sum::<E>()
             })
-            .collect::<Vec<_>>(),
-        2,
-    )
+            .collect::<Vec<E>>();
+
+        RowMajorMatrix::new(
+            (0..target_len)
+                .into_par_iter()
+                .step_by(2)
+                .map(|index| {
+                    let coeff = &folding_coeffs[index >> 1];
+
+                    // 1st part folded with challenge then sum
+                    let cur_same_pos_sum = codeword_fold_with_challenge(
+                        &codeword_to_fold[index..index + 2],
+                        challenge,
+                        *coeff,
+                        inv_2,
+                    );
+                    // 2nd part: retrieve respective index then sum
+                    let next_same_pos_sum = codewords_next_level_matched
+                        .iter()
+                        .map(|codeword| codeword.values[index >> 1])
+                        .sum::<E>();
+                    cur_same_pos_sum + next_same_pos_sum
+                })
+                .collect::<Vec<_>>(),
+            2,
+        )
+    };
+
+    if cfg!(feature = "sanity-check") && is_last_round {
+        let (commitment, merkle_tree) = mmcs_ext.commit_matrix(folded_codeword.clone());
+        commits.push(commitment);
+        trees.push(merkle_tree);
```

Comment:
why did we skip the commit of last round's codeword?

Reply:
add explanation in https://github.com/scroll-tech/ceno/pull/894/commits/3c20a048416b5965f92a4ff88a66ebeca79fbe3e